// @syncfusion/ej2-pdfviewer/src/pdfviewer/print/print.js
import { createElement, Browser } from '@syncfusion/ej2-base';
/**
 * Print module
 */
var Print = /** @class */ (function () {
    /**
     * @private
     */
    function Print(viewer, base) {
        this.pdfViewer = viewer;
        this.pdfViewerBase = base;
    }
    /**
     * Print the PDF document being loaded in the ejPdfViewer control.
     * @returns void
     */
    Print.prototype.print = function () {
        var _this = this;
        var pageIndex;
        if (this.pdfViewerBase.pageCount > 0) {
            // tslint:disable-next-line:max-line-length
            this.printViewerContainer = createElement('div', {
                id: this.pdfViewer.element.id + '_print_viewer_container',
                className: 'e-pv-print-viewer-container'
            });
            this.pdfViewerBase.showPrintLoadingIndicator(true);
            this.iframe = document.createElement('iframe');
            this.iframe.className = 'iframeprint';
            this.iframe.id = 'iframePrint';
            this.iframe.style.position = 'absolute';
            this.iframe.style.top = '-100000000px';
            document.body.appendChild(this.iframe);
            this.frameDoc = this.iframe.contentWindow ? this.iframe.contentWindow : this.iframe.contentDocument;
            this.frameDoc.document.open();
            setTimeout(function () {
                for (pageIndex = 0; pageIndex < _this.pdfViewerBase.pageCount; pageIndex++) {
                    var pageWidth = _this.pdfViewerBase.pageSize[pageIndex].width;
                    var pageHeight = _this.pdfViewerBase.pageSize[pageIndex].height;
                    _this.pdfViewer.printModule.createRequestForPrint(pageIndex, pageWidth, pageHeight, _this.pdfViewerBase.pageCount);
                }
            }, 100);
        }
    };
    Print.prototype.createRequestForPrint = function (pageIndex, pageWidth, pageHeight, pageCount) {
        var proxy = this;
        var request = new XMLHttpRequest();
        // tslint: disable-next-line:max-line-length
        // set default zoomFactor value.  
        var jsonObject = {
            pageNumber: pageIndex, documentId: this.pdfViewerBase.documentId,
            hashId: this.pdfViewerBase.hashId, zoomFactor: 1
        };
        request.open('POST', proxy.pdfViewer.serviceUrl + '/PrintImages', false);
        request.setRequestHeader('Content-Type', 'application/json;charset=UTF-8');
        if (this.pdfViewer.ajaxRequestSettings.ajaxHeaders) {
            this.pdfViewerBase.setCustomAjaxHeaders(request);
        }
        request.send(JSON.stringify(jsonObject));
        // tslint:disable-next-line
        var printImage = request.responseText;
        if (typeof printImage !== 'object') {
            printImage = JSON.parse(printImage);
        }
        var annotationSource = '';
        if (printImage.textMarkupAnnotation && this.pdfViewerBase.isTextMarkupAnnotationModule()) {
            // tslint:disable-next-line:max-line-length
            annotationSource = this.pdfViewer.annotationModule.textMarkupAnnotationModule.printTextMarkupAnnotations(printImage.textMarkupAnnotation, pageIndex);
        }
        // tslint:disable-next-line:max-line-length
        this.printCanvas = createElement('canvas', { id: this.pdfViewer.element.id + '_printCanvas_' + pageIndex, className: 'e-pv-print-canvas' });
        this.printCanvas.style.width = pageWidth + 'px';
        this.printCanvas.style.height = pageHeight + 'px';
        this.printCanvas.height = 1056 * window.devicePixelRatio;
        this.printCanvas.width = 816 * window.devicePixelRatio;
        var context = this.printCanvas.getContext('2d');
        var pageImage = new Image();
        var annotationImage = new Image();
        pageImage.onload = function () {
            if (pageHeight > pageWidth) {
                context.drawImage(pageImage, 0, 0, proxy.printCanvas.width, proxy.printCanvas.height);
                if (annotationSource) {
                    context.drawImage(annotationImage, 0, 0, proxy.printCanvas.width, proxy.printCanvas.height);
                }
            }
            else {
                // translate to center canvas
                context.translate(proxy.printCanvas.width * 0.5, proxy.printCanvas.height * 0.5);
                // rotate the canvas to - 90 degree 
                context.rotate(-0.5 * Math.PI);
                // un translate the canvas back to origin
                context.translate(-proxy.printCanvas.height * 0.5, -proxy.printCanvas.width * 0.5);
                // draw the image
                context.drawImage(pageImage, 0, 0, proxy.printCanvas.height, proxy.printCanvas.width);
                if (annotationSource) {
                    context.drawImage(annotationImage, 0, 0, proxy.printCanvas.height, proxy.printCanvas.width);
                }
            }
            if (pageIndex === (proxy.pdfViewerBase.pageCount - 1)) {
                proxy.printWindowOpen();
            }
        };
        pageImage.src = printImage.image;
        annotationImage.src = annotationSource;
        this.printViewerContainer.appendChild(this.printCanvas);
    };
    Print.prototype.printWindowOpen = function () {
        var _this = this;
        var browserUserAgent = navigator.userAgent;
        // tslint: disable-next-line:max-line-length
        if ((browserUserAgent.indexOf('Chrome') !== -1) || (browserUserAgent.indexOf('Safari') !== -1) ||
            (browserUserAgent.indexOf('Firefox')) !== -1) {
            //chrome and firefox
            this.frameDoc.document.write('<!DOCTYPE html>');
            // tslint: disable-next-line:max-line-length
            this.frameDoc.document.write('<html moznomarginboxes mozdisallowselectionprint><head><style>html, body { height: 100%; }'
                + ' img { height: 100%; width: 100%; display: block; }@media print { body { margin: 0cm; }'
                + ' img { width:100%; max-width: 1048px; box-sizing: border-box; }br, button { display: none; }'
                // set default page Height and page Width for A4 size.
                + ' div{ page-break-inside: avoid; }} @page{margin:0mm; size: 816px 1056px;}</style></head><body><center class="loader">');
        }
        else {
            //ie
            this.frameDoc.document.write('<!DOCTYPE html>');
            // tslint: disable-next-line:max-line-length
            this.frameDoc.document.write('<html><head>'
                + '<style>html, body { height: 99%; } img { height: 99%; width: 100%; }@media print { body { margin: 0cm; }'
                + 'img { width:98%; max-width: 1048px; box-sizing: border-box; }br, button { display: none; } '
                // set default page Height and page Width for A4 size.
                + 'div{ page-break-inside: avoid; }} @page{margin:0mm; size: 816px 1056px;}</style></head><body><center>');
        }
        for (var i = 0; i < this.printViewerContainer.children.length; i++) {
            // tslint:disable-next-line:max-line-length
            var canvasUrl = this.printViewerContainer.children[i].toDataURL();
            this.frameDoc.document.write('<div style="margin:0mm;width:816px;height:1056px;position:relative"><img src="' + canvasUrl + '" id="' + 'image_' + i + '" /></div>');
        }
        this.pdfViewerBase.showPrintLoadingIndicator(false);
        if (Browser.isIE || Browser.info.name === 'edge') {
            try {
                this.iframe.contentWindow.document.execCommand('print', false, null);
            }
            catch (e) {
                this.iframe.contentWindow.print();
            }
        }
        else {
            setTimeout(function () {
                _this.iframe.contentWindow.print();
                _this.iframe.contentWindow.focus();
                document.body.removeChild(_this.iframe);
            }, 200);
        }
    };
    /**
     * @private
     */
    Print.prototype.destroy = function () {
        this.printViewerContainer = undefined;
        this.frameDoc = undefined;
    };
    /**
     * @private
     */
    Print.prototype.getModuleName = function () {
        return 'Print';
    };
    return Print;
}());
export { Print };
