// @syncfusion/ej2-filemanager/src/file-manager/layout/details-view.js
import { Grid, Resize, ContextMenu, Sort, VirtualScroll } from '@syncfusion/ej2-grids';
import { select, KeyboardEvents, EventHandler, getValue, selectAll, isNullOrUndefined } from '@syncfusion/ej2-base';
import { isNullOrUndefined as isNOU, Touch, setValue, addClass, removeClass } from '@syncfusion/ej2-base';
import { Internationalization, closest } from '@syncfusion/ej2-base';
import * as events from '../base/constant';
import * as CLS from '../base/classes';
import { createDialog, createImageDialog } from '../pop-up/dialog';
import { treeNodes } from '../common/utility';
import { removeBlur, openAction, getImageUrl, fileType, getSortedData, getLocaleText } from '../common/utility';
import { createEmptyElement } from '../common/utility';
import { read, Download } from '../common/operations';
import { cutFiles, addBlur, openSearchFolder } from '../common/index';
Grid.Inject(Resize, ContextMenu, Sort, VirtualScroll);
/**
 * GridView module
 */
var DetailsView = /** @class */ (function () {
    /**
     * Constructor for the GridView module
     * @hidden
     */
    function DetailsView(parent) {
        this.isInteracted = true;
        this.pasteOperation = false;
        this.uploadOperation = false;
        this.count = 0;
        this.isRendered = true;
        this.parent = parent;
        this.element = select('#' + this.parent.element.id + CLS.GRID_ID, this.parent.element);
        this.addEventListener();
        this.keyConfigs = {
            altEnter: 'alt+enter',
            esc: 'escape',
            tab: 'tab',
            del: 'delete',
            ctrlX: 'ctrl+x',
            ctrlC: 'ctrl+c',
            ctrlV: 'ctrl+v',
            ctrlShiftN: 'ctrl+shift+n',
            shiftdel: 'shift+delete',
            shiftF10: 'shift+F10',
            f2: 'f2',
            ctrlA: 'ctrl+a',
            enter: 'enter'
        };
    }
    // tslint:disable-next-line
    DetailsView.prototype.render = function (args) {
        if (this.parent.view === 'Details') {
            removeClass([this.parent.element], CLS.MULTI_SELECT);
            var items = getSortedData(this.parent, args.files);
            var columns = this.getColumns();
            var sortSettings = void 0;
            if (this.parent.isMobile) {
                sortSettings = [];
            }
            else {
                sortSettings = [{ direction: this.parent.sortOrder, field: this.parent.sortBy }];
            }
            this.gridObj = new Grid({
                dataSource: items,
                allowSorting: true,
                rowSelected: this.onSelected.bind(this),
                rowDeselected: this.onDeSelection.bind(this),
                allowResizing: this.parent.detailsViewSettings.columnResizing,
                selectionSettings: {
                    type: this.parent.allowMultiSelection ? 'Multiple' : 'Single',
                    checkboxMode: 'ResetOnRowClick'
                },
                enableRtl: this.parent.enableRtl,
                pageSettings: { pageSize: 20 },
                sortSettings: { allowUnsort: false, columns: sortSettings },
                columns: columns,
                recordDoubleClick: this.DblClickEvents.bind(this),
                beforeDataBound: this.onBeforeDataBound.bind(this),
                dataBound: this.onDataBound.bind(this),
                rowDataBound: this.onRowDataBound.bind(this),
                actionBegin: this.onActionBegin.bind(this),
                headerCellInfo: this.onHeaderCellInfo.bind(this),
                width: '100%'
            });
            this.gridObj.appendTo('#' + this.parent.element.id + CLS.GRID_ID);
            this.wireEvents();
            this.adjustHeight();
            // tslint:disable-next-line
            this.gridObj.defaultLocale.EmptyRecord = '';
            this.emptyArgs = args;
        }
    };
    DetailsView.prototype.getColumns = function () {
        var columns;
        if (this.parent.isMobile) {
            columns = [
                {
                    field: 'name', headerText: getLocaleText(this.parent, 'Name'), width: 'auto', minWidth: 120, headerTextAlign: 'Left',
                    template: '<div class="e-fe-text">${name}</div><div class="e-fe-date">${dateModified}</div>' +
                        '<span class="e-fe-size">${size}</span>'
                },
            ];
        }
        else {
            columns = JSON.parse(JSON.stringify(this.parent.detailsViewSettings.columns));
            for (var i = 0, len = columns.length; i < len; i++) {
                columns[i].headerText = getLocaleText(this.parent, columns[i].headerText);
            }
        }
        var iWidth = ((this.parent.isMobile || this.parent.isBigger) ? '54' : '46');
        var icon = {
            field: 'type', width: iWidth, minWidth: iWidth, template: '<span class="e-fe-icon ${iconClass}"></span>',
            allowResizing: false, allowSorting: true, customAttributes: { class: 'e-fe-grid-icon' },
            headerTemplate: '<span class="e-fe-icon e-fe-folder"></span>',
        };
        columns.unshift(icon);
        if (this.parent.allowMultiSelection) {
            var cWidth = (this.parent.isBigger ? '36' : '26');
            var cBox = {
                type: 'checkbox', width: cWidth, minWidth: cWidth, customAttributes: { class: 'e-fe-checkbox' },
                allowResizing: false, allowSorting: false
            };
            if (this.parent.isMobile) {
                columns.push(cBox);
            }
            else {
                columns.unshift(cBox);
            }
        }
        return columns;
    };
    DetailsView.prototype.adjustHeight = function () {
        if (!this.gridObj) {
            return;
        }
        var pane = select('#' + this.parent.element.id + CLS.CONTENT_ID, this.parent.element);
        var bar = select('#' + this.parent.element.id + CLS.BREADCRUMBBAR_ID, this.parent.element);
        var gridHeader = select('.' + CLS.GRID_HEADER, this.parent.element);
        var height = (pane.offsetHeight - bar.offsetHeight - gridHeader.offsetHeight);
        this.gridObj.height = height;
        this.gridObj.dataBind();
    };
    DetailsView.prototype.renderCheckBox = function () {
        this.gridObj.columns = this.getColumns();
        this.gridObj.refreshColumns();
    };
    DetailsView.prototype.onRowDataBound = function (args) {
        /* istanbul ignore next */
        if (!this.parent.showFileExtension && getValue('isFile', args.data)) {
            var textEle = args.row.querySelector('.e-fe-text');
            var name_1 = getValue('name', args.data);
            var type = getValue('type', args.data);
            textEle.innerHTML = name_1.substr(0, name_1.length - type.length);
        }
        if (getValue('size', args.data) !== undefined && args.row.querySelector('.e-fe-size')) {
            var sizeEle = args.row.querySelector('.e-fe-size');
            var modifiedSize = void 0;
            if (!getValue('isFile', args.data)) {
                modifiedSize = '';
            }
            else {
                var sizeValue = getValue('size', args.data);
                if ((sizeValue / 1024) === 0) {
                    modifiedSize = '0 KB';
                }
                else {
                    var intl = new Internationalization();
                    var value = intl.formatNumber((sizeValue / 1024), { format: 'n' });
                    modifiedSize = value + ' KB';
                }
            }
            sizeEle.innerHTML = modifiedSize;
        }
        if (this.parent.isMobile) {
            if (getValue('dateModified', args.data) !== undefined && args.row.querySelector('.e-fe-date')) {
                var dateEle = args.row.querySelector('.e-fe-date');
                var intl = new Internationalization();
                var columns = this.parent.detailsViewSettings.columns;
                var format = void 0;
                for (var i = 0; i < columns.length; i++) {
                    if (columns[i].field === 'dateModified') {
                        format = columns[i].format;
                        break;
                    }
                }
                var formattedString = intl.formatDate(new Date(getValue('dateModified', args.data)), format);
                dateEle.innerHTML = formattedString;
            }
        }
        var checkWrap = args.row.querySelector('.' + CLS.CB_WRAP);
        if (checkWrap) {
            checkWrap.classList.add('e-small');
        }
        var eventArgs = {
            element: args.row,
            fileDetails: args.data,
            module: 'DetailsView'
        };
        this.parent.trigger('beforeFileLoad', eventArgs);
    };
    DetailsView.prototype.onActionBegin = function (args) {
        if (args.requestType === 'sorting') {
            this.parent.sortOrder = args.direction;
            this.parent.sortBy = args.columnName;
            if (this.parent.selectedItems.length !== 0) {
                this.sortItem = true;
                var rows = this.gridObj.getSelectedRowIndexes();
                var len = rows.length;
                this.sortSelectedNodes = [];
                while (len > 0) {
                    var data = this.gridObj.getRowsObject()[rows[len - 1]].data;
                    this.sortSelectedNodes.push(getValue('name', data));
                    len--;
                }
            }
            this.parent.notify(events.sortByChange, {});
        }
    };
    DetailsView.prototype.onHeaderCellInfo = function (args) {
        var checkWrap = args.node.querySelector('.' + CLS.CB_WRAP);
        if (checkWrap) {
            checkWrap.classList.add('e-small');
        }
    };
    DetailsView.prototype.onBeforeDataBound = function (args) {
        var items = getSortedData(this.parent, this.gridObj.dataSource);
        args.result = items;
    };
    DetailsView.prototype.maintainBlur = function () {
        var length = 0;
        var records = this.gridObj.getCurrentViewRecords();
        for (length; length < records.length; length++) {
            var nodeEle = this.gridObj.getDataRows()[length];
            var name_2 = nodeEle.querySelector('.' + CLS.TEMPLATE_CELL).textContent;
            if (this.parent.selectedNodes.indexOf(name_2) !== -1) {
                var node = selectAll('.' + CLS.ROWCELL, nodeEle);
                var nodeLength = 0;
                while (nodeLength < node.length) {
                    addBlur(node[nodeLength]);
                    nodeLength++;
                }
            }
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.onDataBound = function (args) {
        if (this.parent.selectedItems.length !== 0) {
            this.selectedItem = true;
        }
        if (this.pasteOperation === true || this.selectedItem === true) {
            var selectedNodes = (this.selectedItem !== true) ? this.parent.selectedNodes : this.parent.selectedItems;
            this.selectRecords(selectedNodes);
            this.pasteOperation = ((this.selectedItem !== true)) ? false : this.pasteOperation;
            this.selectedItem = false;
        }
        if (this.parent.cutNodes && this.parent.cutNodes.length !== 0) {
            this.maintainBlur();
        }
        if (this.parent.createdItem) {
            this.selectRecords([getValue('name', this.parent.createdItem)]);
            this.parent.createdItem = null;
        }
        if (this.parent.renamedItem) {
            this.selectRecords([getValue('name', this.parent.renamedItem)]);
            this.parent.renamedItem = null;
        }
        if (this.sortItem === true) {
            this.selectRecords(this.sortSelectedNodes);
            this.sortItem = false;
        }
        if (this.parent.allowMultiSelection && this.parent.singleSelection !== undefined) {
            this.selectRecords([this.parent.singleSelection]);
        }
        if (!this.parent.allowMultiSelection && this.parent.singleSelection !== undefined) {
            this.selectRecords([this.parent.singleSelection]);
        }
        if (this.uploadOperation === true) {
            this.count++;
            this.selectRecords(this.parent.uploadItem);
            if (this.count === this.parent.uploadItem.length) {
                this.uploadOperation = false;
                this.parent.uploadItem = [];
            }
        }
        if (this.gridObj.currentViewData.length * this.gridObj.getRowHeight() < this.gridObj.height) {
            var hdTable = this.gridObj.getHeaderContent();
            hdTable.style.paddingRight = '';
            hdTable.style.paddingLeft = '';
            var hdContent = select('.e-headercontent', hdTable);
            hdContent.style.borderRightWidth = '0';
            var cnTable = this.gridObj.getContent().querySelector('.e-content');
            cnTable.style.overflowY = '';
            cnTable.classList.add('e-scrollShow');
        }
        else {
            var hdTable = this.gridObj.getHeaderContent();
            if (!this.parent.enableRtl) {
                hdTable.style.paddingRight = '16px';
            }
            else {
                hdTable.style.paddingLeft = '16px';
            }
            var cnTable = this.gridObj.getContent().querySelector('.e-content');
            cnTable.classList.remove('e-scrollShow');
        }
        this.isRendered = true;
        this.checkEmptyDiv(this.emptyArgs);
    };
    DetailsView.prototype.selectRecords = function (nodes) {
        var gridRecords = this.gridObj.getCurrentViewRecords();
        var sRecords = [];
        for (var i = 0, len = gridRecords.length; i < len; i++) {
            if (nodes.indexOf(getValue('name', gridRecords[i])) !== -1) {
                sRecords.push(i);
            }
        }
        if (sRecords.length !== 0) {
            this.gridObj.selectRows(sRecords);
        }
    };
    DetailsView.prototype.onSortColumn = function (args) {
        this.gridObj.sortModule.sortColumn(this.parent.sortBy, this.parent.sortOrder);
    };
    DetailsView.prototype.onPropertyChanged = function (e) {
        if (e.module !== this.getModuleName() && e.module !== 'common') {
            /* istanbul ignore next */
            return;
        }
        for (var _i = 0, _a = Object.keys(e.newProp); _i < _a.length; _i++) {
            var prop = _a[_i];
            switch (prop) {
                case 'height':
                    this.adjustHeight();
                    break;
                case 'detailsViewSettings':
                    if (!isNullOrUndefined(this.gridObj)) {
                        var columns = this.getColumns();
                        this.gridObj.columns = columns;
                        this.gridObj.allowResizing = this.parent.detailsViewSettings.columnResizing;
                        this.gridObj.dataBind();
                        this.gridObj.refreshColumns();
                    }
                    break;
                case 'selectedItems':
                    if (this.parent.selectedItems.length !== 0) {
                        this.selectRecords(this.parent.selectedItems);
                    }
                    else if (!isNOU(this.gridObj)) {
                        this.gridObj.clearSelection();
                    }
                    break;
                case 'enableRtl':
                    if (!isNullOrUndefined(this.gridObj)) {
                        this.gridObj.enableRtl = e.newProp.enableRtl;
                        this.gridObj.dataBind();
                    }
                    break;
                case 'showFileExtension':
                    read(this.parent, events.pathChanged, this.parent.path);
                    break;
                case 'showHiddenItems':
                    read(this.parent, events.pathChanged, this.parent.path);
                    break;
                case 'allowMultiSelection':
                    if (!isNullOrUndefined(this.gridObj)) {
                        this.gridObj.selectionSettings.type = e.newProp.allowMultiSelection ? 'Multiple' : 'Single';
                        this.renderCheckBox();
                    }
                    break;
                case 'view':
                    read(this.parent, events.layoutChange, this.parent.path);
                    break;
            }
        }
    };
    DetailsView.prototype.onPathChanged = function (args) {
        if (this.parent.breadcrumbbarModule.searchObj.element.value.trim() === '' && this.gridObj) {
            this.parent.searchedItems = [];
            var len = this.gridObj.columns.length;
            // tslint:disable-next-line
            var column = JSON.parse(JSON.stringify(this.gridObj.columns));
            if (column[len - 1].field) {
                if (column[len - 1].field === 'filterPath') {
                    this.gridObj.columns.pop();
                    this.gridObj.refreshColumns();
                }
            }
        }
        removeBlur(this.parent);
        if (!this.parent.persistData) {
            this.parent.setProperties({ selectedItems: [] }, true);
        }
        if (this.parent.view === 'Details') {
            /* istanbul ignore next */
            if (!this.parent.persistData) {
                this.parent.setProperties({ selectedItems: [] }, true);
            }
            else {
                this.isInteracted = false;
            }
            this.parent.persistData = false;
            this.parent.cutNodes = [];
            this.gridObj.dataSource = getSortedData(this.parent, args.files);
            this.parent.notify(events.searchTextChange, args);
        }
        this.emptyArgs = args;
    };
    DetailsView.prototype.checkEmptyDiv = function (args) {
        var items = getSortedData(this.parent, args.files);
        if (items.length === 0 && !isNOU(this.element.querySelector('.' + CLS.GRID_VIEW))) {
            createEmptyElement(this.parent, getValue('name', args), this.element);
        }
        else if (items.length !== 0 && this.element.querySelector('.' + CLS.EMPTY)) {
            if (this.element.querySelector('.' + CLS.GRID_VIEW).querySelector('.' + CLS.EMPTY)) {
                var emptyDiv = this.element.querySelector('.' + CLS.GRID_VIEW).querySelector('.' + CLS.EMPTY);
                this.element.querySelector('.' + CLS.GRID_VIEW).removeChild(emptyDiv);
            }
            else {
                this.element.removeChild(this.element.querySelector('.' + CLS.EMPTY));
            }
        }
    };
    DetailsView.prototype.onOpenInit = function (args) {
        if (this.parent.activeModule === 'detailsview') {
            var data = this.gridObj.getSelectedRecords()[0];
            this.openContent(data);
        }
    };
    /**
     * Triggers when double click on the grid record
     * @public
     */
    DetailsView.prototype.DblClickEvents = function (args) {
        this.gridObj.selectRows([args.rowIndex]);
        var data;
        if (args.rowData) {
            data = JSON.parse(JSON.stringify(args.rowData));
            this.openContent(data);
        }
    };
    DetailsView.prototype.openContent = function (data) {
        var eventArgs = { cancel: false, fileDetails: data };
        this.parent.trigger('beforeFileOpen', eventArgs);
        if (eventArgs.cancel) {
            return;
        }
        if (getValue('isFile', data)) {
            var icon = fileType(data);
            if (icon === CLS.ICON_IMAGE) {
                var name_3 = getValue('name', data);
                var imgUrl = getImageUrl(this.parent, data);
                createImageDialog(this.parent, name_3, imgUrl);
            }
        }
        else {
            var val = this.parent.breadcrumbbarModule.searchObj.element.value;
            if (val === '') {
                var newPath = this.parent.path + getValue('name', data) + '/';
                this.parent.setProperties({ path: newPath }, true);
                this.parent.pathId.push(getValue('nodeId', data));
                this.parent.itemData = [data];
                openAction(this.parent);
            }
            else {
                openSearchFolder(this.parent, data);
            }
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.onLayoutChange = function (args) {
        if (this.parent.view === 'Details') {
            if (!this.gridObj) {
                this.render(args);
            }
            this.gridObj.dataSource = getSortedData(this.parent, args.files);
            this.parent.notify(events.hideLayout, {});
            this.gridObj.element.classList.remove(CLS.DISPLAY_NONE);
            this.isInteracted = false;
            this.gridObj.clearSelection();
            if (this.parent.selectedItems) {
                this.selectedItem = true;
            }
            if (this.parent.breadcrumbbarModule.searchObj.element.value.trim() !== '') {
                this.onSearchFiles(args);
            }
            this.adjustHeight();
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.onSearchFiles = function (args) {
        if (this.parent.view === 'Details') {
            var len = this.gridObj.columns.length;
            // tslint:disable-next-line
            var column = JSON.parse(JSON.stringify(this.gridObj.columns));
            if (column[len - 1].field) {
                if (column[len - 1].field === 'filterPath') {
                    this.gridObj.columns.pop();
                }
            }
            var item = { field: 'filterPath', headerText: 'path', minWidth: 180 };
            if (!this.parent.isMobile) {
                this.gridObj.columns.push(item);
            }
            this.gridObj.refreshColumns();
            this.parent.searchedItems = args.files;
            this.onPathChanged(args);
        }
    };
    DetailsView.prototype.changeData = function (args) {
        this.isInteracted = false;
        this.gridObj.dataSource = getSortedData(this.parent, args.files);
        if (this.parent.selectedItems) {
            this.selectedItem = true;
        }
    };
    DetailsView.prototype.onFinalizeEnd = function (args) {
        if (this.parent.view !== 'Details') {
            return;
        }
        if (!this.gridObj) {
            this.render(args);
            this.parent.notify(events.searchTextChange, args);
        }
        else {
            this.onPathChanged(args);
        }
    };
    DetailsView.prototype.onCreateEnd = function (args) {
        if (this.parent.view !== 'Details') {
            return;
        }
        this.onPathChanged(args);
    };
    DetailsView.prototype.onRenameInit = function () {
        if (this.parent.view === 'Details' && this.parent.selectedItems.length === 1) {
            this.updateRenameData();
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.onDeleteEnd = function (args) {
        if (this.parent.view !== 'Details') {
            return;
        }
        this.onPathChanged(args);
        this.parent.setProperties({ selectedItems: [] }, true);
    };
    DetailsView.prototype.onRefreshEnd = function (args) {
        if (this.parent.view !== 'Details') {
            return;
        }
        this.changeData(args);
    };
    DetailsView.prototype.onHideLayout = function (args) {
        if (this.parent.view !== 'Details' && this.gridObj) {
            this.gridObj.element.classList.add(CLS.DISPLAY_NONE);
        }
    };
    DetailsView.prototype.onSelectAllInit = function () {
        if (this.parent.view === 'Details') {
            this.isInteracted = false;
            this.gridObj.selectionModule.selectRowsByRange(0, this.gridObj.getRows().length);
            this.isInteracted = true;
        }
    };
    DetailsView.prototype.onClearAllInit = function () {
        if (this.parent.view === 'Details') {
            this.removeSelection();
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.onSelectionChanged = function () {
        removeClass([this.element], CLS.HEADER_CHECK);
        if (this.parent.selectedItems.length > 0) {
            addClass([this.element], CLS.HEADER_CHECK);
        }
    };
    DetailsView.prototype.onBeforeRequest = function () {
        this.isRendered = false;
    };
    DetailsView.prototype.onAfterRequest = function (args) {
        this.isRendered = true;
    };
    DetailsView.prototype.addEventListener = function () {
        this.parent.on(events.finalizeEnd, this.onFinalizeEnd, this);
        this.parent.on(events.destroy, this.destroy, this);
        this.parent.on(events.layoutChange, this.onLayoutChange, this);
        this.parent.on(events.pathChanged, this.onPathChanged, this);
        this.parent.on(events.createEnd, this.onCreateEnd, this);
        this.parent.on(events.deleteEnd, this.onDeleteEnd, this);
        this.parent.on(events.refreshEnd, this.onRefreshEnd, this);
        this.parent.on(events.search, this.onSearchFiles, this);
        this.parent.on(events.modelChanged, this.onPropertyChanged, this);
        this.parent.on(events.renameInit, this.onRenameInit, this);
        this.parent.on(events.renameEnd, this.onPathChanged, this);
        this.parent.on(events.openInit, this.onOpenInit, this);
        this.parent.on(events.sortColumn, this.onSortColumn, this);
        this.parent.on(events.openEnd, this.onPathChanged, this);
        this.parent.on(events.hideLayout, this.onHideLayout, this);
        this.parent.on(events.selectAllInit, this.onSelectAllInit, this);
        this.parent.on(events.clearAllInit, this.onClearAllInit, this);
        this.parent.on(events.pathColumn, this.onPathColumn, this);
        this.parent.on(events.selectionChanged, this.onSelectionChanged, this);
        this.parent.on(events.beforeRequest, this.onBeforeRequest, this);
        this.parent.on(events.afterRequest, this.onAfterRequest, this);
    };
    DetailsView.prototype.removeEventListener = function () {
        this.parent.off(events.finalizeEnd, this.onFinalizeEnd);
        this.parent.off(events.destroy, this.destroy);
        this.parent.off(events.layoutChange, this.onLayoutChange);
        this.parent.off(events.pathChanged, this.onPathChanged);
        this.parent.off(events.createEnd, this.onCreateEnd);
        this.parent.off(events.refreshEnd, this.onRefreshEnd);
        this.parent.off(events.search, this.onSearchFiles);
        this.parent.off(events.modelChanged, this.onPropertyChanged);
        this.parent.off(events.renameInit, this.onRenameInit);
        this.parent.off(events.renameEnd, this.onPathChanged);
        this.parent.off(events.openInit, this.onOpenInit);
        this.parent.off(events.sortColumn, this.onSortColumn);
        this.parent.off(events.openEnd, this.onPathChanged);
        this.parent.off(events.hideLayout, this.onHideLayout);
        this.parent.off(events.selectAllInit, this.onSelectAllInit);
        this.parent.off(events.clearAllInit, this.onClearAllInit);
        this.parent.off(events.deleteEnd, this.onDeleteEnd);
        this.parent.off(events.pathColumn, this.onPathColumn);
        this.parent.off(events.selectionChanged, this.onSelectionChanged);
        this.parent.off(events.beforeRequest, this.onBeforeRequest);
        this.parent.off(events.afterRequest, this.onAfterRequest);
    };
    /**
     * For internal use only - Get the module name.
     * @private
     */
    DetailsView.prototype.getModuleName = function () {
        return 'detailsview';
    };
    /**
     * Destroys the GridView module.
     * @method destroy
     * @return {void}
     */
    DetailsView.prototype.destroy = function () {
        if (this.parent.isDestroyed) {
            return;
        }
        this.removeEventListener();
        if (this.gridObj) {
            this.unWireEvents();
            this.gridObj.destroy();
        }
    };
    /**
     * Grid row selected event
     * @private
     */
    /* istanbul ignore next */
    DetailsView.prototype.onSelected = function (args) {
        this.parent.activeElements = this.gridObj.element.querySelectorAll('.' + CLS.ACTIVE);
        this.parent.activeModule = 'detailsview';
        this.selectedRecords();
        this.parent.notify(events.selectionChanged, {});
        this.triggerSelect('select', args);
        if (this.parent.allowMultiSelection) {
            var rows = this.gridObj.getSelectedRowIndexes();
            var len = rows.length;
            if (len > 1) {
                var data = this.gridObj.getRowsObject()[rows[len - 1]].data;
                this.parent.currentItemText = getValue('name', data);
            }
        }
        if (this.parent.selectedItems.length === 1) {
            var data = this.gridObj.getRowsObject()[this.gridObj.selectedRowIndex].data;
            this.parent.currentItemText = getValue('name', data);
        }
        var indexes = getValue('rowIndexes', args);
        var multiSelect = getValue('enableSelectMultiTouch', this.gridObj.selectionModule);
        if (this.parent.isDevice && isNOU(indexes) && args.target && !multiSelect && !args.target.closest('.e-headercell')) {
            this.parent.isFile = getValue('isFile', args.data);
            if (!this.parent.isFile) {
                this.openContent(args.data);
            }
        }
        this.parent.visitedItem = args.row;
    };
    /* istanbul ignore next */
    DetailsView.prototype.onPathColumn = function (args) {
        if (this.parent.view === 'Details') {
            var len = this.gridObj.columns.length;
            if (this.parent.breadcrumbbarModule.searchObj.element.value === '') {
                // tslint:disable-next-line
                var column = JSON.parse(JSON.stringify(this.gridObj.columns));
                if (column[len - 1].field) {
                    if (column[len - 1].field === 'filterPath') {
                        this.gridObj.columns.pop();
                        this.gridObj.refreshColumns();
                    }
                }
            }
        }
    };
    DetailsView.prototype.selectedRecords = function () {
        this.parent.setProperties({ selectedItems: [] }, true);
        var selectedRecords = this.gridSelectNodes();
        var selectSize = 0;
        while (selectSize < selectedRecords.length) {
            var record = selectedRecords[selectSize];
            this.parent.selectedItems.push(record.name);
            selectSize++;
        }
    };
    /**
     * Grid row de-selected event
     * @private
     */
    DetailsView.prototype.onDeSelection = function (args) {
        if (!this.isInteracted) {
            this.isInteracted = true;
            return;
        }
        if (this.parent.activeElements[0].querySelector('.' + CLS.ROWCELL)) {
            this.selectedRecords();
            this.parent.activeElements = this.gridObj.element.querySelectorAll('.' + CLS.ACTIVE);
        }
        var data = args.data;
        for (var i = 0, len = data.length; i < len; i++) {
            var index = this.parent.selectedItems.indexOf(getValue('name', data[i]));
            if (index > -1) {
                this.parent.selectedItems.splice(index, 1);
            }
        }
        if (this.parent.selectedItems.length === 0) {
            setValue('enableSelectMultiTouch', false, this.gridObj.selectionModule);
            removeClass([this.parent.element], CLS.MULTI_SELECT);
        }
        this.parent.notify(events.selectionChanged, {});
        this.triggerSelect('unselect', args);
        this.parent.visitedItem = null;
    };
    DetailsView.prototype.triggerSelect = function (action, args) {
        var eventArgs = { action: action, fileDetails: args.data };
        this.parent.trigger('fileSelect', eventArgs);
    };
    DetailsView.prototype.wireEvents = function () {
        this.wireClickEvent(true);
        this.keyboardModule = new KeyboardEvents(this.gridObj.element, {
            keyAction: this.keyDown.bind(this),
            keyConfigs: this.keyConfigs,
            eventName: 'keyup',
        });
        EventHandler.add(this.gridObj.element, 'focus', this.removeSelection, this);
    };
    DetailsView.prototype.unWireEvents = function () {
        this.wireClickEvent(false);
        this.keyboardModule.destroy();
        EventHandler.remove(this.gridObj.element, 'focus', this.removeSelection);
    };
    DetailsView.prototype.wireClickEvent = function (toBind) {
        if (toBind) {
            var proxy_1 = this;
            var ele = this.gridObj.getContent();
            this.clickObj = new Touch(ele, {
                tapHold: function (e) {
                    if (proxy_1.parent.isDevice) {
                        e.originalEvent.preventDefault();
                        if (proxy_1.parent.allowMultiSelection) {
                            setValue('enableSelectMultiTouch', proxy_1.parent.allowMultiSelection, proxy_1.gridObj.selectionModule);
                            addClass([proxy_1.parent.element], CLS.MULTI_SELECT);
                        }
                        var target = e.originalEvent.target;
                        if (target) {
                            var row = closest(target, '.' + CLS.ROW);
                            var index = proxy_1.gridObj.getRows().indexOf(row);
                            proxy_1.gridObj.selectRow(index);
                        }
                    }
                }
            });
        }
        else {
            if (this.clickObj) {
                this.clickObj.destroy();
            }
        }
    };
    /* istanbul ignore next */
    DetailsView.prototype.removeSelection = function () {
        removeClass([this.parent.element], CLS.MULTI_SELECT);
        this.gridObj.clearSelection();
        this.parent.setProperties({ selectedItems: [] }, true);
        this.parent.notify(events.selectionChanged, {});
    };
    /**
     * Grid keyDown event
     * @private
     */
    /* istanbul ignore next */
    DetailsView.prototype.keyDown = function (e) {
        if (!this.isRendered) {
            return;
        }
        e.preventDefault();
        var action = e.action;
        switch (action) {
            case 'altEnter':
                this.parent.getDetails();
                break;
            case 'esc':
                removeBlur(this.parent);
                this.parent.selectedNodes = [];
                if (this.parent.navigationpaneModule) {
                    this.parent.navigationpaneModule.treeNodes = [];
                }
                break;
            case 'del':
            case 'shiftdel':
                if (this.parent.selectedItems && this.parent.selectedItems.length > 0) {
                    createDialog(this.parent, 'Delete');
                }
                break;
            case 'enter':
                if (this.gridObj.selectedRowIndex === -1) {
                    break;
                }
                var rowData = this.gridObj.getRowsObject()[this.gridObj.selectedRowIndex].data;
                if (rowData) {
                    var data_1 = JSON.parse(JSON.stringify(rowData));
                    this.openContent(data_1);
                }
                break;
            case 'ctrlC':
                removeBlur(this.parent);
                this.parent.navigationpaneModule.treeNodes = [];
                this.parent.navigationpaneModule.copyNodes = [];
                this.parent.cutNodes = [];
                this.parent.selectedNodes = [];
                this.parent.targetPath = this.parent.path;
                treeNodes(this.parent.navigationpaneModule, this.gridSelectNodes(), 'copy');
                this.parent.fileAction = 'CopyTo';
                this.parent.enablePaste = true;
                this.parent.notify(events.showPaste, {});
                this.parent.fileOperation(this.gridSelectNodes());
                break;
            case 'ctrlV':
                this.parent.pasteHandler();
                break;
            case 'ctrlX':
                cutFiles(this.parent);
                this.parent.fileOperation(this.parent.nodeNames);
                break;
            case 'shiftF10':
                Download(this.parent, this.gridSelectNodes());
                break;
            case 'f2':
                if (this.parent.selectedItems.length === 1) {
                    this.updateRenameData();
                    createDialog(this.parent, 'Rename');
                }
                break;
            case 'ctrlA':
                var data = [this.gridObj.getSelectedRecords()[0]];
                this.parent.currentItemText = getValue('name', data[0]);
                break;
            case 'tab':
                var selectedItems = getSortedData(this.parent, this.gridObj.dataSource);
                this.parent.selectedItems = [getValue('name', selectedItems[0])];
                this.selectRecords([getValue('name', selectedItems[0])]);
                break;
        }
    };
    /**
     * Get selected grid records
     * @public
     */
    DetailsView.prototype.gridSelectNodes = function () {
        return this.gridObj.getSelectedRecords();
    };
    DetailsView.prototype.updateRenameData = function () {
        var data = this.gridSelectNodes()[0];
        this.parent.itemData = [data];
        this.parent.currentItemText = getValue('name', data);
        this.parent.isFile = getValue('isFile', data);
    };
    return DetailsView;
}());
export { DetailsView };
