// @syncfusion/ej2-filemanager/src/file-manager/layout/navigation-pane.js
import { TreeView as BaseTreeView } from '@syncfusion/ej2-navigations';
import { isNullOrUndefined as isNOU, select, setValue, getValue } from '@syncfusion/ej2-base';
import { KeyboardEvents, closest } from '@syncfusion/ej2-base';
import { DataManager, Query } from '@syncfusion/ej2-data';
import * as events from '../base/constant';
import * as CLS from '../base/classes';
import { read, Download } from '../common/operations';
import { createDialog } from '../pop-up/dialog';
import { updatePath, getPath, getDirectories } from '../common/utility';
import { removeBlur } from '../common/utility';
import { copyFiles, cutFiles } from '../common/index';
/**
 * `TreeView` module is used to handle Navigation actions.
 */
var NavigationPane = /** @class */ (function () {
    /**
     * Constructor for the TreeView module
     * @hidden
     */
    function NavigationPane(parent) {
        this.treeNodes = [];
        this.removeNodes = [];
        this.expandTree = false;
        this.parent = parent;
        this.addEventListener();
        this.keyConfigs = {
            altEnter: 'alt+enter',
            esc: 'escape',
            del: 'delete',
            ctrlX: 'ctrl+x',
            ctrlC: 'ctrl+c',
            ctrlV: 'ctrl+v',
            ctrlShiftN: 'ctrl+shift+n',
            shiftF10: 'shift+F10',
            f2: 'f2'
        };
    }
    NavigationPane.prototype.onInit = function (args) {
        if (!isNOU(this.treeObj)) {
            return;
        }
        var rootData = getValue('/', this.parent.feParent);
        setValue('icon', 'e-fe-folder', rootData);
        this.rootNode = getValue('name', getValue('/', this.parent.feParent));
        this.treeObj = new BaseTreeView({
            fields: { dataSource: [rootData], id: 'nodeId', text: 'name', hasChildren: 'hasChild', iconCss: 'icon' },
            nodeSelected: this.onNodeSelected.bind(this),
            nodeExpanding: this.onNodeExpand.bind(this),
            allowEditing: true,
            nodeEditing: this.onNodeEditing.bind(this),
            drawNode: this.onDrowNode.bind(this),
            enableRtl: this.parent.enableRtl
        });
        this.treeObj.appendTo('#' + this.parent.element.id + CLS.TREE_ID);
        this.treeObj.element.style.width = '25%';
        this.parent.persistData = true;
        this.wireEvents();
    };
    NavigationPane.prototype.onDrowNode = function (args) {
        var eventArgs = {
            element: args.node,
            fileDetails: args.nodeData,
            module: 'NavigationPane'
        };
        this.parent.trigger('beforeFileLoad', eventArgs);
    };
    NavigationPane.prototype.addChild = function (files, target, prevent) {
        var directories = getDirectories(files);
        if (directories.length > 0) {
            var length_1 = 0;
            var folders = directories;
            while (length_1 < directories.length) {
                folders[length_1].icon = 'e-fe-folder';
                length_1++;
            }
            this.treeObj.addNodes(directories, target, null, prevent);
        }
    };
    /**
     * Tree node selection event
     * @private
     */
    NavigationPane.prototype.onNodeSelected = function (args) {
        if (this.parent.breadcrumbbarModule && this.parent.breadcrumbbarModule.searchObj) {
            this.parent.breadcrumbbarModule.searchObj.element.value = '';
        }
        this.parent.searchedItems = [];
        this.parent.activeElements = this.treeObj.element.querySelectorAll('.' + CLS.ACTIVE);
        if (!args.isInteracted) {
            return;
        }
        var text = getValue('text', args.nodeData);
        this.activeNode = args.node;
        this.parent.activeModule = 'navigationpane';
        if (!this.parent.persistData) {
            this.parent.selectedItems = [];
        }
        updatePath(args.node, text, this.parent);
        this.expandNodeTarget = null;
        if (args.node.querySelector('.' + CLS.ICONS) && args.node.querySelector('.' + CLS.LIST_ITEM) === null) {
            this.expandNodeTarget = 'add';
        }
        this.parent.itemData = this.treeObj.getTreeData(getValue('id', args.nodeData));
        read(this.parent, events.pathChanged, this.parent.path);
        this.parent.visitedItem = args.node;
    };
    /**
     * Tree node expand event
     * @private
     */
    /* istanbul ignore next */
    NavigationPane.prototype.onNodeExpand = function (args) {
        if (!args.isInteracted) {
            return;
        }
        var path = getPath(args.node, getValue('text', args.nodeData));
        if (args.node.querySelector('.' + CLS.LIST_ITEM) === null) {
            this.expandNodeTarget = args.node.getAttribute('data-uid');
            this.parent.expandedId = this.expandNodeTarget;
            this.parent.itemData = this.treeObj.getTreeData(getValue('id', args.nodeData));
            read(this.parent, events.nodeExpand, path);
        }
    };
    /* istanbul ignore next */
    NavigationPane.prototype.onNodeExpanded = function (args) {
        this.addChild(args.files, this.expandNodeTarget, false);
        this.parent.expandedId = null;
    };
    NavigationPane.prototype.onNodeEditing = function (args) {
        if (!isNOU(args.innerHtml)) {
            args.cancel = true;
        }
    };
    NavigationPane.prototype.onPathChanged = function (args) {
        var currFiles = getValue(this.parent.path, this.parent.feFiles);
        if (this.expandNodeTarget === 'add') {
            var sNode = select('[data-uid="' + this.treeObj.selectedNodes[0] + '"]', this.treeObj.element);
            var ul = select('.' + CLS.LIST_PARENT, sNode);
            if (isNOU(ul)) {
                this.addChild(args.files, this.treeObj.selectedNodes[0], true);
            }
            this.expandNodeTarget = '';
        }
        if (isNOU(currFiles)) {
            setValue(this.parent.path, args.files, this.parent.feFiles);
        }
    };
    NavigationPane.prototype.updateTree = function (args) {
        var id = this.treeObj.selectedNodes[0];
        var toExpand = this.treeObj.expandedNodes.indexOf(id) === -1 ? false : true;
        this.removeChildNodes(id);
        setValue(this.parent.path, args.files, this.parent.feFiles);
        this.addChild(args.files, id, !toExpand);
    };
    NavigationPane.prototype.removeChildNodes = function (id) {
        var sNode = select('[data-uid="' + id + '"]', this.treeObj.element);
        var parent = select('.' + CLS.LIST_PARENT, sNode);
        var childs = parent ? Array.prototype.slice.call(parent.children) : null;
        this.treeObj.removeNodes(childs);
    };
    NavigationPane.prototype.onOpenEnd = function (args) {
        var sleId = this.parent.pathId[this.parent.pathId.length - 1];
        this.treeObj.expandAll(this.treeObj.selectedNodes);
        this.treeObj.selectedNodes = [sleId];
        this.expandNodeTarget = 'add';
        this.onPathChanged(args);
    };
    NavigationPane.prototype.onOpenInit = function (args) {
        if (this.parent.activeModule === 'navigationpane') {
            if (args.target.querySelector('.' + CLS.ICONS)) {
                this.treeObj.expandAll(this.treeObj.selectedNodes);
            }
        }
    };
    NavigationPane.prototype.onInitialEnd = function (args) {
        this.onInit(args);
        this.addChild(args.files, getValue('nodeId', args.cwd), false);
    };
    NavigationPane.prototype.onFinalizeEnd = function (args) {
        this.onInit(args);
        var id = getValue('nodeId', args.cwd);
        this.removeChildNodes(id);
        this.addChild(args.files, id, false);
        this.treeObj.selectedNodes = [this.parent.pathId[this.parent.pathId.length - 1]];
    };
    NavigationPane.prototype.onCreateEnd = function (args) {
        this.updateTree(args);
    };
    /* istanbul ignore next */
    NavigationPane.prototype.onDeleteEnd = function (args) {
        if (this.parent.activeModule === 'navigationpane') {
            var selectedNode = this.treeObj.selectedNodes[0];
            var selcetedEle = select('[data-uid="' + selectedNode + '"]', this.treeObj.element);
            var selectedNodeEle = closest(selcetedEle, '.' + CLS.LIST_PARENT).parentElement;
            this.treeObj.selectedNodes = [selectedNodeEle.getAttribute('data-uid')];
            this.treeObj.dataBind();
        }
        this.updateTree(args);
    };
    NavigationPane.prototype.onRefreshEnd = function (args) {
        this.updateTree(args);
    };
    NavigationPane.prototype.onRenameInit = function () {
        if (this.parent.selectedItems.length === 0) {
            this.updateRenameData();
        }
    };
    /* istanbul ignore next */
    NavigationPane.prototype.onRenameEnd = function (args) {
        var resultData = new DataManager(this.treeObj.getTreeData()).
            executeLocal(new Query().where(this.treeObj.fields.text, 'equal', this.parent.currentItemText, false));
        if (resultData[0]) {
            this.treeObj.updateNode(resultData[0][this.treeObj.fields.id].toString(), this.parent.renameText);
        }
    };
    NavigationPane.prototype.onPropertyChanged = function (e) {
        if (e.module !== this.getModuleName() && e.module !== 'common') {
            /* istanbul ignore next */
            return;
        }
        for (var _i = 0, _a = Object.keys(e.newProp); _i < _a.length; _i++) {
            var prop = _a[_i];
            switch (prop) {
                case 'enableRtl':
                    if (this.treeObj) {
                        this.treeObj.enableRtl = e.newProp.enableRtl;
                        this.treeObj.dataBind();
                    }
                    break;
                case 'navigationPaneSettings':
                    read(this.parent, events.finalizeEnd, '/');
                    break;
            }
        }
    };
    /* istanbul ignore next */
    NavigationPane.prototype.onDownLoadInit = function () {
        this.updateActionData();
    };
    NavigationPane.prototype.onSelectionChanged = function (e) {
        this.treeObj.selectedNodes = [e.selectedNode];
    };
    NavigationPane.prototype.onClearPathInit = function (e) {
        this.removeChildNodes(e.selectedNode);
    };
    NavigationPane.prototype.addEventListener = function () {
        this.parent.on(events.modelChanged, this.onPropertyChanged, this);
        this.parent.on(events.downloadInit, this.onDownLoadInit, this);
        this.parent.on(events.initialEnd, this.onInitialEnd, this);
        this.parent.on(events.finalizeEnd, this.onFinalizeEnd, this);
        this.parent.on(events.pathChanged, this.onPathChanged, this);
        this.parent.on(events.nodeExpand, this.onNodeExpanded, this);
        this.parent.on(events.createEnd, this.onCreateEnd, this);
        this.parent.on(events.deleteEnd, this.onDeleteEnd, this);
        this.parent.on(events.refreshEnd, this.onRefreshEnd, this);
        this.parent.on(events.updateTreeSelection, this.onSelectionChanged, this);
        this.parent.on(events.openInit, this.onOpenInit, this);
        this.parent.on(events.openEnd, this.onOpenEnd, this);
        this.parent.on(events.destroy, this.destroy, this);
        this.parent.on(events.renameInit, this.onRenameInit, this);
        this.parent.on(events.renameEnd, this.onRenameEnd, this);
        this.parent.on(events.clearPathInit, this.onClearPathInit, this);
    };
    NavigationPane.prototype.removeEventListener = function () {
        this.parent.off(events.initialEnd, this.onInitialEnd);
        this.parent.off(events.downloadInit, this.onDownLoadInit);
        this.parent.off(events.finalizeEnd, this.onFinalizeEnd);
        this.parent.off(events.modelChanged, this.onPropertyChanged);
        this.parent.off(events.pathChanged, this.onPathChanged);
        this.parent.off(events.updateTreeSelection, this.onSelectionChanged);
        this.parent.off(events.nodeExpand, this.onNodeExpanded);
        this.parent.off(events.createEnd, this.onCreateEnd);
        this.parent.off(events.refreshEnd, this.onRefreshEnd);
        this.parent.off(events.openInit, this.onOpenInit);
        this.parent.off(events.openEnd, this.onOpenEnd);
        this.parent.off(events.destroy, this.destroy);
        this.parent.off(events.renameInit, this.onRenameInit);
        this.parent.off(events.renameEnd, this.onRefreshEnd);
        this.parent.off(events.clearPathInit, this.onClearPathInit);
        this.parent.off(events.deleteEnd, this.onDeleteEnd);
    };
    /**
     * For internal use only - Get the module name.
     * @private
     */
    NavigationPane.prototype.getModuleName = function () {
        return 'navigationpane';
    };
    /**
     * Destroys the TreeView module.
     * @method destroy
     * @return {void}
     */
    NavigationPane.prototype.destroy = function () {
        if (this.parent.isDestroyed) {
            return;
        }
        this.removeEventListener();
        if (this.treeObj) {
            this.unWireEvents();
            this.treeObj.destroy();
        }
    };
    NavigationPane.prototype.wireEvents = function () {
        this.keyboardModule = new KeyboardEvents(this.treeObj.element, {
            keyAction: this.keyDown.bind(this),
            keyConfigs: this.keyConfigs,
            eventName: 'keydown',
        });
    };
    NavigationPane.prototype.unWireEvents = function () {
        this.keyboardModule.destroy();
    };
    /* istanbul ignore next */
    NavigationPane.prototype.keyDown = function (e) {
        var action = e.action;
        var fileObj = this.parent;
        switch (action) {
            case 'altEnter':
                this.parent.getDetails();
                break;
            case 'esc':
                removeBlur(this.parent);
                this.parent.selectedNodes = [];
                this.treeNodes = [];
                break;
            case 'del':
                this.removeNodes = [];
                createDialog(this.parent, 'Delete');
                break;
            case 'ctrlC':
                copyFiles(this.parent);
                fileObj.fileOperation(this.parent.nodeNames);
                this.copyNodes = this.parent.nodeNames;
                break;
            case 'ctrlV':
                fileObj.pasteHandler();
                break;
            case 'ctrlX':
                cutFiles(this.parent);
                fileObj.fileOperation(this.parent.nodeNames);
                break;
            case 'shiftF10':
                Download(this.parent, this.parent.detailsviewModule.gridObj.getSelectedRecords());
                break;
            case 'f2':
                if (this.parent.selectedItems.length === 0) {
                    this.updateRenameData();
                    createDialog(this.parent, 'Rename');
                }
                break;
        }
    };
    NavigationPane.prototype.updateRenameData = function () {
        this.updateActionData();
        this.parent.currentItemText = getValue('name', this.parent.itemData[0]);
    };
    NavigationPane.prototype.updateActionData = function () {
        var data = this.treeObj.getTreeData(this.treeObj.selectedNodes[0])[0];
        this.parent.itemData = [data];
        this.parent.isFile = false;
    };
    /**
     * Move tree folders on cut operation
     * @public
     */
    NavigationPane.prototype.moveNode = function () {
        this.treeObj.moveNodes(this.treeNodes, this.treeObj.selectedNodes[0], null);
        var i = 0;
        var nodes = [];
        var fileObj = this.parent;
        removeBlur(this.parent);
        this.treeNodes = [];
    };
    /**
     * Remove tree folders on delete operation
     * @public
     */
    /* istanbul ignore next */
    NavigationPane.prototype.removeNode = function () {
        this.treeObj.removeNodes(this.removeNodes);
        var fileObj = this.parent;
        removeBlur(this.parent);
        this.removeNodes = [];
    };
    /**
     * Add tree folders on copy operation
     * @public
     */
    NavigationPane.prototype.copyNode = function () {
        this.treeObj.addNodes(this.copyNodes, this.activeNode, null);
        var fileObj = this.parent;
        removeBlur(this.parent);
    };
    return NavigationPane;
}());
export { NavigationPane };
