// @syncfusion/ej2-filemanager/src/file-manager/pop-up/context-menu.js
import { ContextMenu as Menu } from '@syncfusion/ej2-navigations';
import { isNullOrUndefined, createElement, closest, remove, KeyboardEvents } from '@syncfusion/ej2-base';
import { getValue } from '@syncfusion/ej2-base';
import { Download, read } from './../common/operations';
import { createDialog } from './dialog';
import { cutFiles, copyFiles, removeBlur, refresh, getItemObject, getFileObject, getPathObject, getLocaleText } from './../common/utility';
import { getCssClass, sortbyClickHandler } from './../common/utility';
import * as events from './../base/constant';
import * as CLS from '../base/classes';
/**
 * ContextMenu module
 */
var ContextMenu = /** @class */ (function () {
    /**
     * Constructor for the ContextMenu module
     * @hidden
     */
    function ContextMenu(parent) {
        this.parent = parent;
        this.render();
    }
    ContextMenu.prototype.render = function () {
        this.keyConfigs = {
            downarrow: 'downarrow',
            uparrow: 'uparrown'
        };
        this.contextMenu = new Menu({
            enableRtl: this.parent.enableRtl,
            locale: this.parent.locale,
            target: '#' + this.parent.element.id,
            beforeItemRender: this.onBeforeItemRender.bind(this),
            select: this.onSelect.bind(this),
            beforeOpen: this.onBeforeOpen.bind(this),
            cssClass: getCssClass(this.parent, CLS.ROOT_POPUP)
        });
        this.contextMenu.appendTo('#' + this.parent.element.id + CLS.CONTEXT_MENU_ID);
        this.addEventListener();
    };
    /* istanbul ignore next */
    ContextMenu.prototype.onBeforeItemRender = function (args) {
        if (args.item.id === this.getMenuId('largeiconsview')) {
            var iconSpan = createElement('span');
            var element = args.element;
            element.insertBefore(iconSpan, this.parent.view === 'LargeIcons' ? element.childNodes[1] : element.childNodes[0]);
            iconSpan.setAttribute('class', CLS.ICON_LARGE + ' ' + CLS.MENU_ICON);
        }
        if (args.item.id === this.getMenuId('detailsview')) {
            var iconSpan = createElement('span');
            var element = args.element;
            element.insertBefore(iconSpan, this.parent.view === 'Details' ? element.childNodes[1] : element.childNodes[0]);
            iconSpan.setAttribute('class', CLS.ICON_GRID + ' ' + CLS.MENU_ICON);
        }
    };
    /* istanbul ignore next */
    ContextMenu.prototype.onBeforeOpen = function (args) {
        var select = false;
        var uid;
        // tslint:disable-next-line
        var data;
        var treeFolder = false;
        var target = args.event.target;
        if (target.classList.contains('e-spinner-pane')) {
            target = this.parent.navigationpaneModule.activeNode.getElementsByClassName(CLS.FULLROW)[0];
        }
        if (target.classList.contains(CLS.FULLROW)) {
            this.parent.selectedItems.length = 0;
        }
        this.targetElement = this.parent.view === 'Details' ? closest(target, 'tr') : target;
        var view = this.getTargetView(target);
        /* istanbul ignore next */
        if (target.classList.contains(CLS.TREE_VIEW) || closest(target, 'th') ||
            (closest(target, '#' + this.parent.element.id + CLS.BREADCRUMBBAR_ID))) {
            args.cancel = true;
            // tslint:disable-next-line
        }
        else if (!(this.parent.view === 'LargeIcons') && this.targetElement &&
            this.targetElement.classList.contains('e-emptyrow')) {
            this.setLayoutItem(target);
            //Paste
            // this.contextMenu.enableItems([this.getMenuId('Paste')], this.parent.enablePaste, true);
            /* istanbul ignore next */
        }
        else if (closest(target, '.' + CLS.EMPTY)) {
            this.setLayoutItem(target);
            // tslint:disable-next-line
        }
        else if (!target.classList.contains(CLS.MENU_ITEM) && !target.classList.contains(CLS.MENU_ICON) && !target.classList.contains(CLS.SUBMENU_ICON)) {
            /* istanbul ignore next */
            // tslint:disable-next-line
            if (this.parent.view === 'LargeIcons' && !isNullOrUndefined(closest(target, 'li')) && !closest(target, '#' + this.parent.element.id + CLS.TREE_ID)) {
                var eveArgs = { ctrlKey: true, shiftKey: true };
                data = this.parent.visitedData;
                if (!closest(target, 'li').classList.contains('e-active')) {
                    this.parent.largeiconsviewModule.doSelection(target, eveArgs);
                }
                select = true;
            }
            else if (!isNullOrUndefined(closest(target, 'tr'))) {
                uid = this.targetElement.getAttribute('data-uid');
                data = this.parent.detailsviewModule.gridObj.getRowObjectFromUID(uid).data;
                if (isNullOrUndefined(this.targetElement.getAttribute('aria-selected'))) {
                    /* istanbul ignore next */
                    // tslint:disable-next-line
                    this.parent.detailsviewModule.gridObj.selectRows([parseInt(this.targetElement.getAttribute('aria-rowindex'), 10)]);
                }
                select = true;
                /* istanbul ignore next */
            }
            else if (closest(target, '#' + this.parent.element.id + CLS.TREE_ID)) {
                uid = closest(target, 'li').getAttribute('data-uid');
                treeFolder = true;
            }
            /* istanbul ignore next */
            if (select) {
                if (getValue('isFile', data) === true) {
                    this.setFileItem(target);
                }
                else {
                    this.setFolderItem(false);
                }
                /* istanbul ignore next */
            }
            else if (treeFolder) {
                this.setFolderItem(true);
                if (uid === this.parent.pathId[0]) {
                    this.contextMenu.enableItems([this.getMenuId('Delete'), this.getMenuId('Rename')], false, true);
                }
                /* istanbul ignore next */
                // tslint:disable-next-line
            }
            else if (view === 'TreeView' || view === 'GridView' || view === 'LargeIcon') {
                this.setLayoutItem(target);
                //Paste
                // this.contextMenu.enableItems([this.getMenuId('Paste')], this.parent.enablePaste, true);
                /* istanbul ignore next */
            }
            else {
                args.cancel = true;
            }
        }
        if (args.cancel) {
            return;
        }
        this.contextMenu.dataBind();
        var eventArgs = {
            fileDetails: getFileObject(this.parent),
            element: args.element,
            target: target,
            items: this.contextMenu.items,
            menuModule: this.contextMenu,
            cancel: false
        };
        this.parent.trigger('menuOpen', eventArgs);
        args.cancel = eventArgs.cancel;
        if (args.cancel) {
            return;
        }
    };
    /* istanbul ignore next */
    /** @hidden */
    ContextMenu.prototype.getTargetView = function (target) {
        return target.classList.contains(CLS.TREE_VIEW) ?
            'TreeView' : target.classList.contains(CLS.GRID_VIEW) ?
            'GridView' : target.classList.contains(CLS.ICON_VIEW) ?
            'LargeIcon' : target.classList.contains(CLS.LARGE_ICONS) ?
            'LargeIcon' : '';
    };
    ContextMenu.prototype.setFolderItem = function (isTree) {
        this.contextMenu.items = this.getItemData(this.parent.contextMenuSettings.folder);
        this.contextMenu.dataBind();
        if (isTree) {
            this.contextMenu.enableItems([this.getMenuId('Open')], false, true);
        }
        else if (this.parent.selectedItems.length !== 1) {
            this.contextMenu.enableItems([this.getMenuId('Rename')], false, true);
        }
        //Paste
        // this.contextMenu.enableItems([this.getMenuId('Paste')], this.parent.enablePaste, true);
    };
    /* istanbul ignore next */
    ContextMenu.prototype.setFileItem = function (target) {
        this.contextMenu.items = this.getItemData(this.parent.contextMenuSettings.file);
        this.contextMenu.dataBind();
        if (this.parent.selectedItems.length !== 1) {
            this.contextMenu.enableItems([this.getMenuId('Rename')], false, true);
        }
    };
    ContextMenu.prototype.setLayoutItem = function (target) {
        this.contextMenu.items = this.getItemData(this.parent.contextMenuSettings.layout);
        this.contextMenu.dataBind();
        if ((this.parent.view === 'LargeIcons' &&
            (closest(target, '#' + this.parent.element.id + CLS.LARGEICON_ID).getElementsByClassName(CLS.EMPTY).length !== 0))
            || (this.parent.view === 'Details' &&
                (closest(target, '#' + this.parent.element.id + CLS.GRID_ID).getElementsByClassName(CLS.EMPTY).length !== 0))) {
            this.contextMenu.enableItems([this.getMenuId('SelectAll')], false, true);
            this.contextMenu.dataBind();
        }
    };
    /* istanbul ignore next */
    ContextMenu.prototype.onSelect = function (args) {
        if (isNullOrUndefined(args.item) || !args.item.id) {
            return;
        }
        var itemText = args.item.id.substr((this.parent.element.id + '_cm_').length);
        var details;
        if (itemText === 'refresh' || itemText === 'newfolder') {
            details = getPathObject(this.parent);
            this.parent.itemData = [details];
        }
        else {
            details = getFileObject(this.parent);
        }
        var eventArgs = {
            cancel: false,
            element: args.element,
            fileDetails: details,
            item: args.item
        };
        this.parent.trigger('menuClick', eventArgs);
        if (eventArgs.cancel) {
            return;
        }
        if (this.parent.selectedItems.length > 0 && ((itemText === 'paste') ||
            (itemText === 'upload'))) {
            var path = void 0;
            // tslint:disable-next-line
            var data = void 0;
            if (this.parent.view === 'Details') {
                var uid = this.targetElement.getAttribute('data-uid');
                data = this.parent.detailsviewModule.gridObj.getRowObjectFromUID(uid).data;
                /* istanbul ignore next */
            }
            else {
                var elements = this.targetElement.parentElement;
                data = getItemObject(this.parent, elements);
            }
            /* istanbul ignore next */
            if (data.isFile) {
                path = '';
            }
            else {
                path = data.name + '/';
                this.parent.navigationpaneModule.treeObj.selectedNodes = [data.id];
            }
            var newPath = this.parent.path + path;
            this.parent.setProperties({ path: newPath }, true);
        }
        // tslint:disable-next-line
        var items = this.parent.selectedItems;
        switch (itemText) {
            case 'cut':
                cutFiles(this.parent);
                /* istanbul ignore next */
                if (this.parent.nodeNames) {
                    this.parent.fileOperation(this.parent.nodeNames);
                }
                break;
            case 'copy':
                copyFiles(this.parent);
                /* istanbul ignore next */
                if (this.parent.nodeNames) {
                    this.parent.fileOperation(this.parent.nodeNames);
                }
                /* istanbul ignore next */
                if (this.parent.activeModule === 'navigationpane') {
                    // tslint:disable-next-line
                    this.parent.navigationpaneModule.copyNodes = this.parent.nodeNames;
                }
                break;
            case 'paste':
                this.parent.pasteHandler();
                removeBlur(this.parent);
                break;
            case 'delete':
                createDialog(this.parent, 'Delete');
                break;
            /* istanbul ignore next */
            case 'download':
                /* istanbul ignore next */
                if (this.parent.activeModule === 'detailsview') {
                    items = this.parent.detailsviewModule.gridObj.getSelectedRecords();
                }
                else if (this.parent.activeModule === 'largeiconsview') {
                    var elements = this.parent.activeElements;
                    for (var ele = 0; ele < elements.length; ele++) {
                        items[ele] = getItemObject(this.parent, elements[ele]);
                    }
                }
                else if (this.parent.activeModule === 'navigationpane' && this.parent.selectedItems.length === 0) {
                    this.parent.notify(events.downloadInit, {});
                    items = this.parent.itemData;
                }
                if (items.length > 0) {
                    Download(this.parent, items);
                }
                break;
            case 'rename':
                this.parent.notify(events.renameInit, {});
                createDialog(this.parent, 'Rename');
                break;
            case 'selectall':
                /* istanbul ignore next */
                this.parent.notify(events.selectAllInit, {});
                break;
            case 'refresh':
                refresh(this.parent);
                break;
            case 'open':
                if (this.parent.visitedItem) {
                    this.parent.notify(events.openInit, { target: this.parent.visitedItem });
                }
                break;
            case 'details':
                this.parent.getDetails();
                break;
            case 'newfolder':
                createDialog(this.parent, 'NewFolder');
                break;
            case 'upload':
                document.getElementById(this.parent.element.id + '_upload').click();
                break;
            /* istanbul ignore next */
            case 'name':
            /* istanbul ignore next */
            case 'size':
            /* istanbul ignore next */
            case 'date':
            /* istanbul ignore next */
            case 'ascending':
            /* istanbul ignore next */
            case 'descending':
                /* istanbul ignore next */
                sortbyClickHandler(this.parent, args);
                break;
            // tslint:disable-next-line
            /* istanbul ignore next */
            case 'largeiconsview':
                this.changeLayout('LargeIcons');
                break;
            // tslint:disable-next-line
            /* istanbul ignore next */
            case 'detailsview':
                this.changeLayout('Details');
                break;
        }
    };
    /* istanbul ignore next */
    ContextMenu.prototype.changeLayout = function (view) {
        if (this.parent.view !== view) {
            this.parent.setProperties({ view: view }, true);
            read(this.parent, events.layoutChange, this.parent.path);
        }
    };
    ContextMenu.prototype.onPropertyChanged = function (e) {
        if (e.module !== this.getModuleName() && e.module !== 'common') {
            /* istanbul ignore next */
            return;
        }
        for (var _i = 0, _a = Object.keys(e.newProp); _i < _a.length; _i++) {
            var prop = _a[_i];
            switch (prop) {
                case 'cssClass':
                    this.contextMenu.cssClass = getCssClass(this.parent, CLS.ROOT_POPUP);
                    break;
                case 'enableRtl':
                    this.contextMenu.enableRtl = e.newProp.enableRtl;
                    break;
            }
        }
    };
    ContextMenu.prototype.addEventListener = function () {
        this.parent.on(events.destroy, this.destroy, this);
        this.parent.on(events.modelChanged, this.onPropertyChanged, this);
        this.keyboardModule = new KeyboardEvents(this.contextMenu.element, {
            keyAction: this.keyActionHandler.bind(this),
            keyConfigs: this.keyConfigs,
            eventName: 'keydown',
        });
    };
    ContextMenu.prototype.removeEventListener = function () {
        this.parent.off(events.destroy, this.destroy);
        this.parent.off(events.modelChanged, this.onPropertyChanged);
        this.keyboardModule.destroy();
    };
    ContextMenu.prototype.keyActionHandler = function (e) {
        switch (e.action) {
            case 'uparrow':
            case 'downarrow':
                e.preventDefault();
        }
    };
    /**
     * For internal use only - Get the module name.
     * @private
     */
    ContextMenu.prototype.getModuleName = function () {
        return 'contextmenu';
    };
    /**
     * Destroys the ContextMenu module.
     * @method destroy
     * @return {void}
     */
    ContextMenu.prototype.destroy = function () {
        if (this.parent.isDestroyed) {
            return;
        }
        this.removeEventListener();
        this.contextMenu.destroy();
        if (document.getElementById(this.parent.element.id + CLS.CONTEXT_MENU_ID)) {
            remove(document.getElementById(this.parent.element.id + CLS.CONTEXT_MENU_ID));
        }
    };
    /* istanbul ignore next */
    ContextMenu.prototype.getItemData = function (data) {
        var items = [];
        for (var i = 0; i < data.length; i++) {
            var item = void 0;
            var itemId = this.getMenuId(data[i]);
            var itemText = getLocaleText(this.parent, data[i]);
            switch (data[i]) {
                case '|':
                    item = { separator: true };
                    break;
                case 'Open':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_OPEN };
                    break;
                case 'Upload':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_UPLOAD };
                    break;
                case 'Cut':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_CUT };
                    break;
                case 'Copy':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_COPY };
                    break;
                case 'Paste':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_PASTE };
                    break;
                case 'Delete':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_DELETE };
                    break;
                case 'Rename':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_RENAME };
                    break;
                case 'NewFolder':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_NEWFOLDER };
                    break;
                case 'Details':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_DETAILS };
                    break;
                case 'SortBy':
                    item = {
                        id: itemId, text: itemText, iconCss: CLS.ICON_SHORTBY,
                        items: [
                            {
                                id: this.getMenuId('Name'), text: getLocaleText(this.parent, 'Name'),
                                iconCss: this.parent.sortBy === 'name' ? CLS.TB_OPTION_DOT : null
                            },
                            {
                                id: this.getMenuId('Size'), text: getLocaleText(this.parent, 'Size'),
                                iconCss: this.parent.sortBy === 'size' ? CLS.TB_OPTION_DOT : null
                            },
                            {
                                id: this.getMenuId('Date'), text: getLocaleText(this.parent, 'DateModified'),
                                iconCss: this.parent.sortBy === 'dateModified' ? CLS.TB_OPTION_DOT : null
                            },
                            { separator: true },
                            {
                                id: this.getMenuId('Ascending'), text: getLocaleText(this.parent, 'Ascending'),
                                iconCss: this.parent.sortOrder === 'Ascending' ? CLS.TB_OPTION_TICK : null
                            },
                            {
                                id: this.getMenuId('Descending'), text: getLocaleText(this.parent, 'Descending'),
                                iconCss: this.parent.sortOrder === 'Descending' ? CLS.TB_OPTION_TICK : null
                            }
                        ]
                    };
                    break;
                /* istanbul ignore next */
                case 'View':
                    item = {
                        id: itemId, text: itemText, iconCss: this.parent.view === 'Details' ? CLS.ICON_GRID : CLS.ICON_LARGE,
                        items: [
                            {
                                id: this.getMenuId('largeiconsview'), text: getLocaleText(this.parent, 'View-LargeIcons'),
                                iconCss: this.parent.view === 'Details' ? null : CLS.TB_OPTION_TICK
                            },
                            {
                                id: this.getMenuId('detailsview'), text: getLocaleText(this.parent, 'View-Details'),
                                iconCss: this.parent.view === 'Details' ? CLS.TB_OPTION_TICK : null
                            }
                        ]
                    };
                    break;
                case 'Refresh':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_REFRESH };
                    break;
                case 'SelectAll':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_SELECTALL };
                    break;
                /* istanbul ignore next */
                case 'Download':
                    item = { id: itemId, text: itemText, iconCss: CLS.ICON_DOWNLOAD };
                    break;
                /* istanbul ignore next */
                default:
                    item = { id: itemId, text: itemText };
                    break;
            }
            items.push(item);
        }
        return items;
    };
    ContextMenu.prototype.getMenuId = function (id) {
        return this.parent.element.id + '_cm_' + id.toLowerCase();
    };
    return ContextMenu;
}());
export { ContextMenu };
